File: java/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonTriangulation/monotoneDecomposition/_RandomSegmentOrder.java
package vsdk.toolkit.environment.geometry.geometricProcessing.polygonTriangulation.monotoneDecomposition;

import java.util.ArrayList;

public final class _RandomSegmentOrder {
    private static int nextPermutationIndex;
    private static ArrayList<Integer> segmentPermutation = new ArrayList<>();
    private static long seed = 1L;

    private _RandomSegmentOrder() {}

    private static double randomUnit() {
        // Park-Miller "minimal standard" rand() sequence, matching the C++ runtime here.
        seed = (16807L * seed) % 2147483647L;
        return (double) seed / (double) 2147483647L;
    }

    public static int generateRandomOrdering(int n) {
        nextPermutationIndex = 1;
        segmentPermutation.clear();
        for (int i = 0; i <= n; i++) segmentPermutation.add(0);

        ArrayList<Integer> localPermutation = new ArrayList<>();
        for (int i = 0; i <= n; i++) localPermutation.add(i);

        int pBase = 0;
        for (int i = 1; i <= n; i++, pBase++) {
            int m = ((int)Math.floor(randomUnit() * 32000.0)) % (n + 1 - i) + 1;
            segmentPermutation.set(i, localPermutation.get(pBase + m));
            if (m != 1) {
                localPermutation.set(pBase + m, localPermutation.get(pBase + 1));
            }
        }
        return 0;
    }

    public static int chooseSegment() {
        return segmentPermutation.get(nextPermutationIndex++);
    }
}

File: cpp/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonTriangulation/monotoneDecomposition/_RandomSegmentOrder.cpp
#include <cmath>

#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/geometry/geometricProcessing/polygonTriangulation/monotoneDecomposition/_RandomSegmentOrder.h"

// References: [SEID1991] Seidel, R. "A simple and Fast Randomized Algorithm
// for Computing Trapezoidal Decompositions and for Triangulating Polygons".

int _RandomSegmentOrder::nextPermutationIndex;
java::ArrayList<int> _RandomSegmentOrder::segmentPermutation;

double _RandomSegmentOrder::randomUnit() {
    return ((double)rand() / (double)RAND_MAX);
}

int _RandomSegmentOrder::generateRandomOrdering(int n) {
    // [SEID1991].3 Random segmentPermutation for incremental insertion order.
    int i;
    int m;
    java::ArrayList<int> localPermutation;
    int *p;

    nextPermutationIndex = 1;
    segmentPermutation.clear();
    segmentPermutation.reserve(n + 1);
    for (i = 0; i <= n; i++) {
        segmentPermutation.add(0);
    }

    for (i = 0; i <= n; i++) {
        localPermutation.add(i);
    }

    p = localPermutation.data();
    for (i = 1; i <= n; i++, p++) {
        m = (int)(floor((randomUnit() * 32000))) % (n + 1 - i) + 1;
        segmentPermutation[i] = p[m];
        if (m != 1) {
            p[m] = p[1];
        }
    }
    return 0;
}

int _RandomSegmentOrder::chooseSegment(void) {
    return segmentPermutation[nextPermutationIndex++];
}
